Do not use $_POST directly in validate_checkout

AlexP11223 opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

During checkout validation, there are points where we specifically rely on various `$_POST` superglobal values being set, which made reuse of our validation functionality difficult. This change tidies this up.

Closes #35328.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

Checkout validation should continue to work as it did previously. More specifically, we should ensure that we are still validating if the custom has agreed to the shop's terms and conditions (since the change touches that check):

1. Ensure you have designated a terms and conditions page (via **WooCommerce ▸ Settings ▸ Advanced ▸ Page Setup**):

<img width="800" alt="terms-and-conditions" src="https://user-images.githubusercontent.com/3594411/198156779-a3cae656-de7c-481e-ae03-15d1d8f1e20c.png">

2. As a customer, add a product to your cart and attempt to checkout. In the first instance, populate the checkout form correctly but avoid checking the box that indicates you agree with the terms and conditions:

<img width="300" src="https://user-images.githubusercontent.com/3594411/198157195-8ccc1c90-6e42-40ad-bf59-d967a434a94d.png" />

3. Consequently, on trying to checkout, you should see a warning as follows:

![tc-error](https://user-images.githubusercontent.com/3594411/198157251-a741ea19-5a37-4b9b-9600-acd2acb0c5f8.png)

4. Rinse and repeat; except agree to the terms and conditions. If all other fields are correctly populated, you should be able to checkout successfully.
5. Repeat again, but try to supply invalid values to other checkout fields (for example, you might provide an invalid postal code) and confirm validation successfully prevents this. 

<!-- End testing instructions -->

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

AlexP11223 committed:
Do not use $_POST directly in validate_checkout

barryhughes committed:
Changelog.

barryhughes reviewed (dismissed):
Looks good, thanks! We can merge once checks pass (unfortunately, a glitch meant our coding standards check was initially failing for reasons unrelated to your change).

barryhughes committed:
Address some of the undocumented hook errors flagged by our linting check.

This does not address all of the undocumented hooks (there are many in this file), just the ones flagged by the GitHub Actions code sniff check. This is an effort to unblock the PR, which we otherwise cannot merge at present.

barryhughes committed:
Add missing `@since` tags (satisfy linter).

barryhughes commented:
Hitting pause on this review (internal convo: p1666827490378459-slack-C03CPM3UXDJ). 

It seems that to unblock the required check we will ultimately need to document all the currently undocumented hooks, though they are otherwise nothing to do with the core change this PR was raised to address (and I'm reluctant to push placeholder text to satisfy the linter, since that defeats the purpose).
